### ISSUE TITLE
proc-macro: Couldn't parse default expression `Trait`

### ISSUE THREAD

**Issue by saruman9** [rust-analyzer/rust-analyzer#7810]
## MWE

`Cargo.toml`:

```toml
[dependencies]
derive_builder = "0.9.0"
```

`main.rs`:

```rust
use derive_builder::Builder;
fn main() {}

#[derive(Debug, Default, Builder)]
#[builder(default)]
struct Test {}

```

`cargo check` reports that everything is ok:

```shell
$ cargo check 2>/dev/null && echo $?
0
```

`rust-analyzer` (VSCode) with `rust-analyzer.cargo.loadOutDirsFromCheck` and `rust-analyzer.procMacro.enable` options enabled reports:
```
proc macro returned error: proc-macro panicked: Couldn't parse default expression `Trait`: "expected curly braces"
```
and

```shell
$ rust-analyzer diagnostics --load-output-dirs --with-proc-macro ./
...
Diagnostic { message: "proc macro process is closed", range: 43..112, severity: Error, fix: None, unused: false, code: Some(DiagnosticCode("macro-error")) }

diagnostic scan complete

[ERROR rust_analyzer] Unexpected error: diagnostic error detected
diagnostic error detected

```

## Versions

- Rust version: `1.50.0 (cb75ad5db 2021-02-10)` (stable)
- rust-analyzer version: 14de9e5

**Comment by jonas-schievink** [comment]
This seems to happen here: https://github.com/colin-kiegel/rust-derive-builder/blob/c72cd242b282d12dcc55dcb8d3e322830c26eec2/derive_builder_core/src/block.rs#L59

...where the input string is `::std::default::Default::default()`. Given that there's a `format!` that will produce `{::std::default::Default::default()}`, I'm not sure why syn complains about missing curly braces?

**Comment by kevinmehall** [comment]
The `proc macro process is closed` from the CLI is unrelated, and it's not even running the proc-macro there. I opened #7843 for that one.

**Comment by kevinmehall** [comment]
The issue here is that [r-a's implementation of `TokenStream`](https://github.com/rust-analyzer/rust-analyzer/blob/25a43e24c81f7b51234fc4e65e5e36ddbe3d60f1/crates/proc_macro_srv/src/rustc_server.rs#L27) holds a `tt::Subtree` and assumes its `delimiter` is `None`. The iterator implementation iterates over the inner `token_trees` and ignores the delimiter.  However, `TokenStream::from_str` violates this assumption when the input consists of a single `Group` by producing a `Subtree` with an outer delimiter, which gets ignored.

An even more minimal example:
```
#[proc_macro]
pub fn test(_input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let ts = "(x)".parse::<proc_macro::TokenStream>().unwrap();
    dbg!(ts.to_string(), ts.clone().into_iter().collect::<Vec<proc_macro::TokenTree>>());
    ts
}
```

produces in r-a:
```
[test-proc-macro/src/lib.rs:4] ts.to_string() = "(x)"
[test-proc-macro/src/lib.rs:4] ts.clone().into_iter().collect::<Vec<proc_macro::TokenTree>>() = [
    Ident {
        ident: "x",
        span: 4294967295,
    },
]
```

vs in rustc:
```
[test-proc-macro/src/lib.rs:4] ts.to_string() = "(x)"
[test-proc-macro/src/lib.rs:4] ts.clone().into_iter().collect::<Vec<proc_macro::TokenTree>>() = [
    Group {
        delimiter: Parenthesis,
        stream: TokenStream [
            Ident {
                ident: "x",
                span: #7 bytes(279..303),
            },
        ],
        span: #7 bytes(279..303),
    },
]
```

I think the fix is to make `TokenStream` contain `Vec<tt::TokenTree>` instead of `tt::Subtree` to encode that assumption in the type system instead of having a delimiter field that is partially ignored. When converting a `tt:Subtree` to a `TokenStream`, it should only be flattened if the `delimiter` is `None`, otherwise left as-is. If that sounds right, I'll try making a PR.

**Comment by edwin0cheng** [comment]
@kevinmehall How about we just fix `TokenStream::from_str` such that it wrapped a Delimited::None Subtree ?

**Comment by kevinmehall** [comment]
Yeah, that would be the minimally invasive fix. `mbe::parse_to_token_tree(src)` already returns a `delimiter: None` subtree for input that does not consist of a single Group, so it would have to check for that before wrapping it again, or it would expose an extra level of Group to proc macros.

The reason I suggested the `TokenTree` change is that `Subtree` is just `pub delimiter: Option<Delimiter>, pub token_trees: Vec<TokenTree>`, so a `Subtree` that is statically guaranteed not to have a delimiter is just `Vec<TokenTree>`.

**Comment by edwin0cheng** [comment]
> he reason I suggested the TokenTree change is that Subtree is just pub delimiter: Option<Delimiter>, pub token_trees: Vec<TokenTree>, so a Subtree that is statically guaranteed not to have a delimiter is just Vec<TokenTree>.

Sound reasonable !
